### PR TITLE
Work out better merge heuristics & uniqueness test

### PR DESCRIPTION
**Pull request by natschil**
Handles merge heuristics better, and checks the returned vortices for uniqueness.

### CODE REVIEW

**Comment by natschil** [comment]
If all tests pass, we could merge this IMO.

**Comment by dkarrasch** [comment]
Are you trying to buy the reviewer? 🤣

File: examples/ocean_flow.jl [hunk 1]
```diff
@@ -108,7 +108,7 @@ P = AxisArray(SVector{2}.(xspan, yspan'), xspan, yspan)
 # with OECSs, we enable the combine_31 heuristic
 
 S = map(rate_of_strain_tensor, P)
-p = LCSParameters(boxradius=2.5, pmin=-1, pmax=1,combine_31=true)
+p = LCSParameters(boxradius=2.5, pmin=-1, pmax=1,merge_heuristics=[combine_20,combine_31])
```
Comment (dkarrasch) [inline examples/ocean_flow.jl]: spaces after commas

File: src/ellipticLCS.jl [hunk 2]
```diff
@@ -76,8 +76,10 @@ Container for parameters used in elliptic LCS computations.
 ## Fields
 * `boxradius`: "radius" of localization square for closed orbit detection
 * `indexradius=1e-1boxradius`: radius for singularity type detection
-* `combine_pairs=true`: whether isolated singularity pairs should be merged
-* `combine_31=true` : whether 1 trisector + 2 wedge configurations should be merged
+* `merge_heuristics`: a list of heuristics for combining singularities, supported are
+* * `combine_20` merge isolated singularity pairs that are mutually nearest neighbors
+* * `combine_31` : merge 1 trisector + nearest-neighbor 3 wedge configurations.
+* * `combine_20_aggressive`: an additional wedge combination heuristic
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: consistency for colons, spaces, etc.

File: src/ellipticLCS.jl [hunk 3]
```diff
@@ -116,8 +117,7 @@ struct LCSParameters
     function LCSParameters(
                 boxradius::Real,
                 indexradius::Real=1e-1boxradius,
-                combine_pairs::Bool=true,
-                combine_31::Bool=false,
+                merge_heuristics=[combine_20],
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: The default above is to have both heuristics?

File: src/ellipticLCS.jl [hunk 4]
```diff
@@ -140,8 +140,7 @@ end
 function LCSParameters(;
             boxradius::Real=1.0,
             indexradius::Real=1e-1boxradius,
-            combine_pairs::Bool=true,
-            combine_31::Bool=false,
+            merge_heuristics=[combine_20],
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: Same here. We need consistent defaults.

File: src/ellipticLCS.jl [hunk 5]
```diff
@@ -345,6 +344,10 @@ end
 """
     combine_31(singularities)
 
+Takes the list of singularities in `singularities` and combines them so that any -1/2 singularity whose three nearest
+neighbors are 1/2 singularities becomes an elliptic region, provided that the -1/2 singularity is in the triangle spanned
+by the wedges. This configuration is common for OECS, applying to material
+barriers on a large turbulet example yielded only about an additional 1% material barriers.
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: Could you break the lines here earlier so that they fit into the usual editor window?

File: src/ellipticLCS.jl [hunk 6]
```diff
@@ -393,59 +396,128 @@ function combine_31_configuration(singularities::Vector{Singularity{T}}) where {
...
+            push!(new_singularities, singularities[i])
+        end
+    end
+    return new_singularities
+end
+
+
 """
-    critical_point_detection(vs, combine_distance, dist=s1dist; combine_pairs=true,combine_31=true) -> Vector{Singularity}
+    critical_point_detection(vs, combine_distance, dist=s1dist; merge_heuristics=[combine_20]) -> Vector{Singularity}
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: defaults...

File: src/ellipticLCS.jl [hunk 7]
```diff
@@ -393,59 +396,128 @@ function combine_31_configuration(singularities::Vector{Singularity{T}}) where {
...
+Heuristics listed in `merge_heuristics` cf. [`LCSParams`](@ref) are applied to combine singularities.
 
 Returns a vector of [`Singularity`](@ref)s.
 """
 function critical_point_detection(vs::AxisArray{<: SVector{2,<:Real},2},
                                     combine_distance::Real,
                                     dist::Function=s1dist;
-                                    combine_pairs=true,
-                                    combine_31=true)
+                                    merge_heuristics=[combine_20]
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: defaults...

File: src/ellipticLCS.jl [hunk 8]
```diff
@@ -393,59 +396,128 @@ function combine_31_configuration(singularities::Vector{Singularity{T}}) where {
...
-        new_singularities =  combine_31_configuration(new_singularities)
+    for f in merge_heuristics
+        new_singularities = f(singularities)
     end
     return new_singularities
 end
 
 """
-    singularity_detection(T, combine_distance; combine_isolated_wedges=true) -> Vector{Singularity}
+    singularity_detection(T, combine_distance; combine_20=true) -> Vector{Singularity}
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: `combine_20` should be `merge_heuristics`, if I'm not mistaken.

File: src/ellipticLCS.jl [hunk 9]
```diff
@@ -652,21 +724,37 @@ function ellipticLCS(T::AbstractMatrix{SymmetricTensor{2,2,S,3}},
...
-    singularities = singularity_detection(T, p.indexradius; combine_pairs=p.combine_pairs,combine_31=p.combine_31)
+    singularities = singularity_detection(T, p.indexradius; merge_heuristics=p.merge_heuristics)
     if singularity_predicate != nothing
         singularities = filter(singularity_predicate,singularities)
     end
     verbose && @info "Found $(length(singularities)) singularities..."
     vortexcenters = singularities[getindices(singularities) .== 1]
     verbose && @info "Defined $(length(vortexcenters)) Poincaré sections..."
 
+    vortices = findVortices(T,vortexcenters,p; verbose=verbose, kwargs...)
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: spaces after commas

File: src/ellipticLCS.jl [hunk 10]
```diff
@@ -652,21 +724,37 @@ function ellipticLCS(T::AbstractMatrix{SymmetricTensor{2,2,S,3}},
...
+    if unique_vortices
+        vortices = makeVortexListUnique(vortices, p.indexradius)
+    end
+
+    return vortices,singularities
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: space

File: src/ellipticLCS.jl [hunk 11]
```diff
@@ -652,21 +724,37 @@ function ellipticLCS(T::AbstractMatrix{SymmetricTensor{2,2,S,3}},
...
-                        singularity_predicate=nothing) where S <: Real
+                        unique_vortices=true,
+                        singularity_predicate=nothing,
+                        kwargs...) where S <: Real
     # detect centers of elliptic (in the index sense) regions
-    xspan = T.axes[1]
-    xmax = xspan[end]
-    singularities = singularity_detection(T, p.indexradius; combine_pairs=p.combine_pairs,combine_31=p.combine_31)
+    singularities = singularity_detection(T, p.indexradius; merge_heuristics=p.merge_heuristics)
     if singularity_predicate != nothing
```
Comment (dkarrasch) [inline src/ellipticLCS.jl]: Checking for `nothing` is done via `=== nothing`, checking for not-nothing is via `!== nothing`.

File: src/exports.jl [hunk 12]
```diff
@@ -23,6 +23,9 @@ export
 	ellipticLCS,
 	constrainedLCS,
 	materialBarriers,
+    combine_20,
+    combine_31,
+    combine_20_aggressive,
```
Comment (dkarrasch) [inline src/exports.jl]: double-check that indention is correct.

File: src/ellipticLCS.jl (hunk 3 above)
Comment (natschil) [reply src/ellipticLCS.jl]: No default is to have just combine_20

File: src/ellipticLCS.jl (hunk 4 above)
Comment (natschil) [reply src/ellipticLCS.jl]: yes this is all consistent

File: src/ellipticLCS.jl (hunk 3 above)
Comment (dkarrasch) [reply src/ellipticLCS.jl]: okay, then in line 111 we need to remove `combine_31`?

Comment (dkarrasch) [reply src/ellipticLCS.jl]: Ah, sorry, I messed it up. I overlooked that the above is from the ocean_example, where we can, of course, overwrite the defaults.

File: test/test_elliptic.jl [hunk 13]
```diff
@@ -13,7 +13,7 @@ const CS = CoherentStructures
             @test length(S) == 1
             @test iszero(S[1].coords)
             @test S[1].index == 1
-            S = @inferred critical_point_detection(v, 0.1; combine_pairs=false)
+            S = @inferred critical_point_detection(v, 0.1; Any[])
```
Comment (dkarrasch) [inline test/test_elliptic.jl]: This should be `merge_heuristics=Any[]`, I assume. Same in line 26.

**Pull request merged**